### ISSUE TITLE
Investigate SQLite database locked errors

### ISSUE THREAD

**Issue by natemcmaster** [dotnet/efcore#2356]
SQLITE_LOCKED errors occurs intermittently between runs. Example output:

> ```
>  Microsoft.Data.Entity.Update.DbUpdateException : An error occurred while updating the entries. See the inner exception for details.
>   ---- Microsoft.Data.Sqlite.SqliteException : Error: database is locked
> ```

See full stack trace: https://gist.github.com/natemcmaster/0536e259226f884f59e7

**Comment by natemcmaster** [comment]
Occurs under these conditions:

1. Shared-cache mode is used by multiple connections
2. Multiple connections attempt to read/write the same table.

Solutions:

1. Don't use shared-cache mode (unfortunately we need this.)
2. https://www.sqlite.org/unlock_notify.html. This article shows a method for setting up blocking calls instead of failing when SQLITE_LOCKED is returned.
3. (For our testing projects), don't run tests in parallel.

Recommend closing this bug because fixing the problem requires calls to native SQLite methods.

**Comment by bricelam** [comment]
Does it only occur when the connections are on different threads?

It's a fair amount of work, but I really like the idea of blocking (with a timeout) in the ADO.NET layer. This would make it transparently "just work" for consumers.

**Comment by natemcmaster** [comment]
Yes, it can only happen when using different threads. Moving to https://github.com/aspnet/Microsoft.Data.Sqlite/issues/97

**Comment by bricelam** [comment]
Hmm, if it's only cross-thread, we should be able to prevent it by using one database per class/collection.

**Comment by natemcmaster** [comment]
Unfortunately, I don't think it will be that easy. The errors I found in EF testing originate from looping over SqliteDataReader.Read().

**Comment by RonFrick** [comment]
Anyone else seeing this after upgrading to rc2. I am getting this a lot while running our sync code. 

Microsoft.Data.Sqlite.SqliteException : Error: database is locked

**Comment by natemcmaster** [comment]
@RonFrick if you can create a minimal repro, feel free to open a new issue.

**Comment by dsdwilches** [comment]
@natemcmaster @RonFrick Did you solve your problem? I'm having the same issue and it's not evident what in my code can be causing this error.

**Comment by dwilches** [comment]
For other people coming here, this guy made a minimal repro in this other issue: https://github.com/aspnet/EntityFrameworkCore/issues/9994
